### ISSUE TITLE
`dat.json` structure proposals

### ISSUE THREAD

**Issue by chartgerink** [p2pcommons/specs#8]
The current module specification (`v0.1.5`) specifies the following structure for the `dat.json` file containing all the metadata (example for `type: content`):

```js
{
  "title": "testing update",
  "description": "fkdslajflkdsj",
  "url": "00a4f2f18bb6cb4e9ba7c2c047c8560d34047457500e415d535de0526c6b4f23",
  "type": "content",
  "subtype": "content",
  "main": "016224398501000307.pdf",
  "license": "https://creativecommons.org/publicdomain/zero/1.0/legalcode",
  "authors": [],
  "parents": []
}
```

As an outsider, it is currently unknown why this format is used or where it comes from. We (@juliangruber and @dpaez + me) have been discussing this in the [`@p2pcommons/working-group`](https://github.com/p2pcommons/working-group/blob/master/20191105.md) and I will make a few proposals here to the `module` specification. Feel free to voice your concerns and favorites :)

# Proposal 1

* Add a `p2pcommons` name/value pair with a string describing the spec used *

```js
{
  "title": "testing update",
  "description": "fkdslajflkdsj",
  "url": "00a4f2f18bb6cb4e9ba7c2c047c8560d34047457500e415d535de0526c6b4f23",
  "type": "content",
  "subtype": "content",
  "main": "016224398501000307.pdf",
  "license": "https://creativecommons.org/publicdomain/zero/1.0/legalcode",
  "authors": [],
  "parents": [],
  "p2pcommons": "Module specification vX.X.X"
}
```

# Proposal 2

* Add a `p2pcommons` name/value pair with a URL referring to the spec *

```js
{
  "title": "testing update",
  "description": "fkdslajflkdsj",
  "url": "00a4f2f18bb6cb4e9ba7c2c047c8560d34047457500e415d535de0526c6b4f23",
  "type": "content",
  "subtype": "content",
  "main": "016224398501000307.pdf",
  "license": "https://creativecommons.org/publicdomain/zero/1.0/legalcode",
  "authors": [],
  "parents": [],
  "p2pcommons": "https://p2pcommons/specs/module/x.x.x"
}
```

# Proposal 3

* Encapsulate all `p2pcommons` specific name/value pairs in a `p2pcommons` name/value pair containing a JSON object * (a bit like the [`dat.json` spec](https://github.com/datprotocol/dat.json))

```js
{
  "title": "testing",
  "description": "flkjdsafjdsf",
  "url": "00a4f2f18bb6cb4e9ba7c2c047c8560d34047457500e415d535de0526c6b4f23",
  "license": "https://creativecommons.org/publicdomain/zero/1.0/legalcode",
  "p2pcommons": {
    "type": "content",
    "subtype": "qXXXXX",
    "main": "XXX.pdf",
    "authors": [],
    "parents": []
  }
}
```

# Proposal 4

* Add a link object to the specification in accordance with [`dat.json` spec](https://github.com/datprotocol/dat.json) *

```js
{
  "title": "testing update",
  "description": "fkdslajflkdsj",
  "url": "00a4f2f18bb6cb4e9ba7c2c047c8560d34047457500e415d535de0526c6b4f23",
  "type": "content",
  "subtype": "content",
  "main": "016224398501000307.pdf",
  "links": [
     "license": "https://creativecommons.org/publicdomain/zero/1.0/legalcode",
     "p2pcommons": "https://p2pcommons/specs/module/x.x.x"
  ],
  "authors": [],
  "parents": [],
}
```

# Proposal 5

* A combination of proposals 3 + 4 *

```js
{
  "title": "testing",
  "description": "flkjdsafjdsf",
  "url": "00a4f2f18bb6cb4e9ba7c2c047c8560d34047457500e415d535de0526c6b4f23",
  "links": [
     "license": "https://creativecommons.org/publicdomain/zero/1.0/legalcode",
     "p2pcommons": "https://p2pcommons/specs/module/x.x.x"
  ],
  "p2pcommons": {
    "type": "content",
    "subtype": "qXXXXX",
    "main": "XXX.pdf",
    "authors": [],
    "parents": []
  }
}
```

**Comment by juliangruber** [comment]
For reference, here the last attempt at creating a spec for the `dat.json`: https://github.com/datprotocol/dat.json

**Comment by juliangruber** [comment]
> * Add a link object to the specification in accordance with [`dat.json` spec](https://github.com/datprotocol/dat.json) *
 
Corrected syntax:

> ```js
> {
>   "title": "testing update",
>   "description": "fkdslajflkdsj",
>   "url": "00a4f2f18bb6cb4e9ba7c2c047c8560d34047457500e415d535de0526c6b4f23",
>   "type": "content",
>   "subtype": "content",
>   "main": "016224398501000307.pdf",
>   "links": {
>      "license": [{"href": "https://creativecommons.org/publicdomain/zero/1.0/legalcode"}],
>      "spec": [{"href": "https://p2pcommons/specs/module/x.x.x"}]
>   },
>   "authors": [],
>   "parents": [],
> }
> ```

**Comment by juliangruber** [comment]
> # Proposal 3
> * Encapsulate all `p2pcommons` specific name/value pairs in a `p2pcommons` name/value pair containing a JSON object * (a bit like the [`dat.json` spec](https://github.com/datprotocol/dat.json))
> 
> ```js
> {
>   "title": "testing",
>   "description": "flkjdsafjdsf",
>   "url": "00a4f2f18bb6cb4e9ba7c2c047c8560d34047457500e415d535de0526c6b4f23",
>   "license": "https://creativecommons.org/publicdomain/zero/1.0/legalcode",
>   "p2pcommons": {
>     "type": "content",
>     "subtype": "qXXXXX",
>     "main": "XXX.pdf",
>     "authors": [],
>     "parents": []
>   }
> }
> ```

I like this the most! Because it cleanly encapsulates all our special properties, and helps prevent naming conflicts between various specs, if that ever becomes a situation.

**Comment by chartgerink** [comment]
I think we settled on this pretty much after the call (https://github.com/p2pcommons/working-group/pull/5).

```js
{
  "title": "testing",
  "description": "flkjdsafjdsf",
  "url": "00a4f2f18bb6cb4e9ba7c2c047c8560d34047457500e415d535de0526c6b4f23",
  "links": {
     "license": [{"href": "https://creativecommons.org/publicdomain/zero/1.0/legalcode"}],
     "spec": [{"href": "https://p2pcommons/specs/module/x.x.x"}]
  },,
  "p2pcommons": {
    "type": "content",
    "subtype": "qXXXXX",
    "main": "XXX.pdf",
    "authors": [],
    "parents": []
  }
}
```

- [ ] I will doublecheck with the [weblinks spec](https://html.spec.whatwg.org/multipage/semantics.html#the-link-element) before implementing it in the spec.

**Comment by chartgerink** [comment]
Closing this as it's being addressed in the #8 pull request.
